pkg/utils/tasks/tasks.go

```diff
@@ -31,64 +35,233 @@ type TaskManager struct {
+func (tm *TaskManager) Queue(task Task, opts ...Option) error {
+	return tm.QueueContext(context.Background(), task, opts...)
+}
+
 // Queue a task with the specified context. Blocks if the queue is full.
-func (tm *TaskManager) QueueContext(ctx context.Context, task Task) {
-	tm.queue <- &TaskHandler{task, ctx}
+func (tm *TaskManager) QueueContext(ctx context.Context, task Task, opts ...Option) error {
+	options := makeOptions(opts...)
+	handler := &TaskHandler{
+		parent:    tm,
+		task:      task,
+		opts:      options,
+		ctx:       ctx,
+		err:       &Error{err: options.err},
+		attempts:  0,
+		retryAt:   time.Time{},
+		scheduled: time.Now(),
+	}
+	return tm.queueTask(handler)
 }
 
-// Queue a task with a background context. Blocks if the queue is full.
-func (tm *TaskManager) Queue(task Task) {
-	tm.QueueContext(context.Background(), task)
+// Queue a task defined by the handler to the specified channel. If the task returns an
+// error, this method queues the task for retry by sending it to the scheduler.
+func (tm *TaskManager) queueTask(handler *TaskHandler) error {
+	// The read-lock allows us to check tm.stopped concurrently. If Stop() has been
+	// called it holds a write lock that prevents this lock from being acquired until
+	// the scheduler has closed the queue.
+	tm.RLock()
+	defer tm.RUnlock()
+
+	if tm.stopped {
+		// Dropping the task because the task manager is not running
+		log.Warn().Err(ErrTaskManagerStopped).Msg("cannot queue async task when task manager is stopped")
+		return ErrTaskManagerStopped
+	}
+
+	// Queue the handler
+	tm.queue <- handler
+	return nil
+}
+
+type TaskHandler struct {
+	parent    *TaskManager
+	task      Task
+	opts      *options
+	ctx       context.Context
+	err       *Error
+	attempts  int
+	retryAt   time.Time
+	scheduled time.Time
 }
 
-func TaskWorker(wg *sync.WaitGroup, queue <-chan *TaskHandler) {
+// Execute the wrapped task with the context. If the task fails, schedule the task to
+// be retried using the backoff specified in the options.
+func (h *TaskHandler) Exec() {
+	// Attempt to execute the task
+	var err error
+	if err = h.task.Do(h.ctx); err == nil {
+		// Success!
+		log.Debug().
+			Dur("duration", time.Since(h.scheduled)).
+			Int("attempts", h.attempts).
+			Msg("async tasks completed")
+		return
+	}
+
+	// Deal with the error
+	h.attempts++
+	h.err.Append(err)
+
+	// Check if we have retries left
+	if h.attempts <= h.opts.retries {
+		// Schedule the retry be added back to the queue
+		h.retryAt = time.Now().Add(h.opts.backoff.NextBackOff())
+		log.Warn().
+			Err(err).
+			Time("retry_at", h.retryAt).
+			Int("attempts", h.attempts).
+			Int("retries_remaining", h.opts.retries-h.attempts).
+			Msg("async task failed, retrying")
+
+		h.parent.queueTask(h)
+		return
+	}
+
+	// At this point we've exhausted all possible retries, so log the error.
+	h.err.Since(h.scheduled)
+	h.err.Log(log.Logger)
+	h.err.Capture(sentry.GetHubFromContext(h.ctx))
+}
```

Comment:
I moved the closure you created into a method on Exec and added a little logging.